nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/status/history/VolatileComponentStatusRepository.java

```diff
@@ -164,6 +183,182 @@ public StatusHistory getRemoteProcessGroupStatusHistory(final String remoteGroup
         return getStatusHistory(remoteGroupId, true, DEFAULT_RPG_METRICS, start, end, preferredDataPoints);
     }
 
+    @Override
+    public StatusHistory getNodeStatusHistory() {
+        final List<NodeStatus> nodeStatusList = nodeStatuses.asList();
+        final List<List<GarbageCollectionStatus>> gcStatusList = gcStatuses.asList();
+        final LinkedList<StatusSnapshot> snapshots = new LinkedList<>();
+
+        final Set<MetricDescriptor<?>> metricDescriptors = new HashSet<>(DEFAULT_NODE_METRICS);
+        final List<MetricDescriptor<List<GarbageCollectionStatus>>> gcMetricDescriptors = new LinkedList<>();
+        final List<MetricDescriptor<List<GarbageCollectionStatus>>> gcMetricDescriptorsDifferential = new LinkedList<>();
+        final List<MetricDescriptor<NodeStatus>> contentStorageStatusDescriptors = new LinkedList<>();
+        final List<MetricDescriptor<NodeStatus>> provenanceStorageStatusDescriptors = new LinkedList<>();
+
+        int ordinal = DEFAULT_NODE_METRICS.size() - 1;
+
+        // Uses the first measurement as reference for repository metrics descriptors
+        if (nodeStatusList.size() > 0) {
+            final NodeStatus referenceNodeStatus = nodeStatusList.get(0);
+            int contentStorageNumber = 0;
+            int provenanceStorageNumber = 0;
+
+            for (int i = 0; i < referenceNodeStatus.getContentRepositories().size(); i++) {
+                final int storageNumber = i;
+                final int counter = metricDescriptors.size() - 1 + NUMBER_OF_STORAGE_METRICS * contentStorageNumber;
+
+                contentStorageStatusDescriptors.add(new StandardMetricDescriptor<>(
```

Comment:
Could use
```suggestion
                metricDescriptors.add(new StandardMetricDescriptor<NodeStatus>(
```
With this approach we could get rid of all the intermediary lists.

Reply:
I reworked this part a bit. Mainly simplification. Also I extracted the descriptor creation parts to make the flow easier to read.